File: DriftGame/uracer-desktop/src/com/bitfire/postprocessing/filters/Lens.java
package com.bitfire.postprocessing.filters;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.bitfire.utils.ShaderLoader;

/**
 * Lens flare effect.
 **/
public final class Lens extends Filter<Lens> {
    private final Vector2 lightPosition = new Vector2();
    private float intensity;
    private final Vector3 color = new Vector3();
    private final Vector2 viewport = new Vector2();

    public Lens(float width, float height) {
        super(ShaderLoader.INSTANCE.fromFile("screenspace", "lensflare"));
        viewport.set(width, height);
        rebind();
    }

    public float getIntensity() {
        return intensity;
    }

    public void setIntensity(float intensity) {
        this.intensity = intensity;
        setParam(Param.Intensity, intensity);
    }

    public Vector3 getColor() {
        return color;
    }

    public void setColor(float r, float g, float b) {
        color.set(r, g, b);
        setParam(Param.Color, color);
    }

    @Override
    public void rebind() {
        // Re-implement super to batch every parameter
        setParams(Param.Texture, u_texture0);
        setParams(Param.LightPosition, lightPosition);
        setParams(Param.Intensity, intensity);
        setParams(Param.Color, color);
        setParams(Param.Viewport, viewport);
        endParams();
    }

    @Override
    protected void onBeforeRender() {
        inputTexture.bind(u_texture0);
    }

    public enum Param implements Parameter {

        Texture("u_texture0", 0), LightPosition("u_lightPosition", 2), Intensity("u_intensity", 0), Color("u_color", 3), Viewport(
			"u_viewport", 2);


        private final String mnemonic;
        private final int elementSize;

        Param(String mnemonic, int arrayElementSize) {
            this.mnemonic = mnemonic;
            this.elementSize = arrayElementSize;
        }

        @Override
        public String mnemonic() {
            return this.mnemonic;
        }

        @Override
        public int arrayElementSize() {
            return this.elementSize;
        }
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/postprocessing/filters/Lens.kt
package com.bitfire.postprocessing.filters

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.math.Vector2
import com.bitfire.utils.ShaderLoader

/**
 * Lens flare is a graphical effect that simulates the way that light scatters and reflects within a camera lens, creating visual
 * artifacts like glows, streaks, and rings around bright light sources.
 *
 * This class implements the lens flare effect as described in John Chapman's article (without lens dirt or diffraction starburst). Lens color image
 * (lenscolor.png) is located in src/main/resources/ folder.
 *
 * @see [http://john-chapman-graphics.blogspot.co.uk/2013/02/pseudo-lens-flare.html](http://john-chapman-graphics.blogspot.co.uk/2013/02/pseudo-lens-flare.html)
 */
class Lens(width: Int, height: Int) : Filter<Lens>(ShaderLoader.fromFile("screenspace", "lensflare2")) {

    private val viewportInverse = Vector2(1F / width, 1F / height)
    private var ghosts = 0
    private var haloWidth = 0F
    private var lensColorTexture: Texture? = null

    init {
        rebind()
    }

    fun setGhosts(ghosts: Int) {
        this.ghosts = ghosts
        setParam(Param.Ghosts, ghosts)
    }

    fun setHaloWidth(haloWidth: Float) {
        this.haloWidth = haloWidth
        setParam(Param.HaloWidth, haloWidth)
    }

    fun setLensColorTexture(tex: Texture) {
        this.lensColorTexture = tex
        setParam(Param.LensColor, u_texture1)
    }

    override fun rebind() {
        // Re-implement super to batch every parameter
        setParams(Param.Texture, u_texture0)
        setParams(Param.LensColor, u_texture1)
        setParams(Param.ViewportInverse, viewportInverse)
        setParams(Param.Ghosts, ghosts)
        setParams(Param.HaloWidth, haloWidth)
        endParams()
    }

    override fun onBeforeRender() {
        inputTexture.bind(u_texture0)
        lensColorTexture!!.bind(u_texture1)
    }

    enum class Param(private val mnemonic: String, private val arrayElementSize: Int) : Parameter {
        Texture("u_texture0", 0),
        LensColor("u_texture1", 0),
        ViewportInverse("u_viewportInverse", 2),
        Ghosts("u_ghosts", 0),
        HaloWidth("u_haloWidth", 0);

        override fun mnemonic() = this.mnemonic
        override fun arrayElementSize() = this.arrayElementSize
    }
}
